bundle/edu.gemini.sp.vcs/src/main/scala/edu/gemini/sp/vcs/diff/VcsServer.scala

```diff
@@ -0,0 +1,169 @@
+        } else {
+          VcsAction(a)
+        }
+      }
+    }
+
+  /** Adds the given program to the database, provided it doesn't share the
+    * same key or id with an existing program in the database and the user has
+    * appropriate keys.
+    */
+  // TODO: seems to work but can probably be clened up a bit
+  def add(p: ISPProgram, user: Set[Principal]): VcsAction[Unit] =
+    Option(p.getProgramID).toRightDisjunction(MissingId).liftVcs >>= { id =>
+      val progKey = p.getProgramKey
+      try {
+        writeLock(progKey)
+        accessControlled(id, user) {
+          Option(odb.lookupProgramByID(id)).fold {
+            Option(odb.lookupProgram(p.getProgramKey)).fold {
+              putProg(p).liftVcs
+            } { _ => VcsAction.fail(KeyAlreadyExists(id, progKey)) }
+          } { _ => VcsAction.fail(IdAlreadyExists(id)) }
+        }
+      } finally {
+        writeUnlock(progKey)
+      }
+    }
+
+  /** Server implementation of `VcsService`. */
+  final class SecureVcsService(user: Set[Principal]) extends VcsService {
+    def geminiPrincipals: Set[GeminiPrincipal] =
+      user.collect { case p: GeminiPrincipal => p }
+
+    override def version(id: SPProgramID): TryVcs[VersionMap] =
+      vs.read(id, user)(_.getVersions).unsafeRun
+
+    override def add(p: ISPProgram): TryVcs[Unit] =
+      vs.add(p, user).unsafeRun
+
+    override def checkout(id: SPProgramID): TryVcs[ISPProgram] =
+      vs.read(id, user)(identity).unsafeRun
+
+    override def diffState(id: SPProgramID): TryVcs[DiffState] =
+      vs.read(id, user)(DiffState.apply).unsafeRun
+
+    override def fetchDiffs(id: SPProgramID, state: DiffState): TryVcs[MergePlan.Transport] =
+      vs.read(id, user) { p =>
+        vcsLog.log(OpFetch, id, geminiPrincipals)
+        ProgramDiff.compare(p, state)
+      }.map(_.encode).unsafeRun
+
+    override def storeDiffs(id: SPProgramID, mpt: MergePlan.Transport): TryVcs[Boolean] = {
+      val mp = mpt.decode
+      vs.write[Boolean](id, user,
+        p => VersionMap.compare(mp.vm(p), p.getVersions) match {
+          case Newer => VcsAction(true)
+          case Same  => VcsAction(false)
+          case _     => VcsAction.fail(NeedsUpdate)
+        },
+        identity,
+        (f, p, _) =>
+          for {
+            _ <- mp.merge(f, p)
+            _ <- VcsAction(vcsLog.log(OpStore, id, geminiPrincipals))
+          } yield ()
+      ).unsafeRun
+    }
+
+    override def log(id: SPProgramID, offset:Int, length:Int): TryVcs[(List[VcsEventSet], Boolean)] =
+      try {
+        vcsLog.selectByProgram(id, offset, length).right
+      } catch {
+        case ex: Exception => VcsException(ex).left
+      }
+  }
+
+  private def locking[A](id: SPProgramID, user: Set[Principal], lock: SPNodeKey => Unit, unlock: SPNodeKey => Unit)(body: ISPProgram => VcsAction[A]): VcsAction[A] =
+    accessControlled(id, user) {
+      lookup(id) >>= { p =>
```

Comment:
`>>=` comes from Haskell and is basically `flatMap` but it reads better.  Also goes well with `>>` which is a `flatMap` in which the argument is ignored: `m.flatMap { _ => ... }`.